lib/rspec/core/memoized_helpers.rb

```diff
@@ -123,14 +123,40 @@ def is_expected
 
       # @private
       def __memoized
-        @__memoized ||= {}
+        @__memoized ||= Memoized.new
+      end
+
+      # @private
+      class Memoized
+        def initialize
+          @memoized = {}
+          @mutex    = Mutex.new
+        end
+
+        def for(key, &initializer)
+          return @memoized[key][0] if @memoized.key? key
+
+          value = initializer.call
```

Comment:
It looks to me like `initializer` could get called multiple times.  IMO this has the potential to cause buggy behavior:

* If multiple threads get here, it's non-deterministic which value will actually be memoized and returned by the `let`-defined method -- it's basically whichever the value for whichever thread gets into the `synchronize` block first, which we have no control over.
* Besides computing a value or creating an object, the `let` definition may cause some side effect, and having it run multiple times could be very problematic.  Consider, for example, a typical rails spec that creates an AR record using `let` and then has later test logic that assumes that record is the only one in the DB table.  If `initializer.call` is executed twice by two different threads, the DB table could have 2 records instead of 1 and the user would get intermittent, non-deterministic test failures.